syntax = "proto3";

option go_package = ".;proto";

import "google/protobuf/empty.proto";

service Goods {
    // banner
    rpc GetBannerList(BannerFilterRequest)returns(BannerListResponse);
    rpc CreateBanner(BannerRequest)returns(BannerInfoResponse);
    rpc UpdateBanner(BannerRequest)returns(google.protobuf.Empty);
    rpc DeleteBanner(BannerByIdRequest)returns(google.protobuf.Empty);

    // 分类
    rpc CategoryList(CategoryFilterRequest)returns(CategoryListResponse);
    rpc CreateCategory(CategoryRequest)returns(CategoryInfoResponse);
    rpc DeleteCategory(CategoryByIdRequest)returns(google.protobuf.Empty);
    rpc UpdateCategory(CategoryRequest)returns(google.protobuf.Empty);

    // 品牌
    rpc GetBrandList(BrandFilterRquest)returns(BrandListResponse);
    rpc CreateBrand(CreateBrandRequest)returns(BrandInfoResponse);
    rpc DeleteBrand(BrandByIdRequest)returns(google.protobuf.Empty);
    rpc UpdateBrand(BrandInfoRequest)returns(google.protobuf.Empty);

    // 获取指定 category 下 brand列表
    rpc GetBrandsByCategory(CategoryByIdRequest)returns(BrandListResponse);

    // 获取所有分类品牌 = 类名+品牌名
    rpc CategoryBrandList(CategoryBrandFilterRequest)returns(CategoryBrandListResponse);
    rpc UpdateCategoryBrand(CategoryBrandRequest)returns(CategoryBrandResponse);
    rpc CreateCategoryBrand(CategoryBrandRequest)returns(CategoryBrandResponse);
    rpc DeleteCategoryBrand(CategoryBrandRequest)returns(google.protobuf.Empty);

    // 商品
    rpc GoodsList(GoodsFilterRequest)returns(GoodsListResponse);
    rpc BatchGetGoods(BatchGoodsByIdRequest)returns(GoodsListResponse);
    rpc GetGoodsDetail(GoodsByIdRequest)returns(GoodsDetailResponse);
    rpc CreateGoods(GoodsRequest)returns(GoodsDetailResponse);
    rpc DeleteGoods(GoodsByIdRequest)returns(google.protobuf.Empty);
    rpc UpdateGoods(GoodsRequest)returns(google.protobuf.Empty);
    rpc UpdateStatus(GoodsRequest)returns (google.protobuf.Empty);
}

// banner
message BannerFilterRequest {
    sint32 page = 1;
    sint32 limit = 2;
}

message BannerListResponse {
    sint32 total = 1;
    repeated BannerInfoResponse data = 2;
}

message BannerRequest {
    int32 id = 1;
    int32 index = 2;
    string image = 3;
    string url = 4;
}

message BannerInfoResponse {
    sint32 id = 1;
    int32 index = 2;
    string image = 3;
    string url = 4;
}

message BannerByIdRequest {
    sint32 id =1;
}

// 分类
message CategoryFilterRequest {
    sint32 id = 1;
    sint32 parent_id = 2;
}

message CategoryListResponse {
    int32 total = 1;
    repeated CategoryInfoResponse data = 2;
    string JsonData = 3;
}

message CategoryRequest {
    sint32 id = 1;
    string name = 2;
    sint32 parent_id = 3;
    bool is_tab = 4;
    int32 level = 5;
}

message CategoryInfoResponse {
    sint32 id = 1;
    string name = 2;
    sint32 parent_id = 3;
    sint32 level = 4;
    bool is_tab = 5;
}

message CategoryByIdRequest {
    sint32 id = 1;
}

// 品牌
message BrandFilterRquest {
    sint32 page = 1;
    sint32 limit =2;
}

message BrandListResponse {
    int32 total = 1;
    repeated BrandInfoResponse data = 2;
}

message CreateBrandRequest {
    string name = 1;
    string logo = 2;
}

message BrandInfoResponse {
    sint32 id = 1;
    string name = 2;
    string logo = 3;
}

message BrandByIdRequest {
    sint32 id = 1;
}

message BrandInfoRequest {
    sint32 id = 1;
    string name = 2;
    string logo = 3;
}

// 分类商品
message CategoryBrandFilterRequest {
    int32 page =  1;
    int32 limit = 2;
}

message CategoryBrandRequest {
    sint32 id = 1;
    sint32 category_id = 2;
    sint32 brand_id = 3;
}

message CategoryBrandListResponse {
    sint32 total = 1;
    repeated CategoryBrandResponse data = 2;
}

message CategoryBrandResponse {
    sint32 id = 1;
    BrandInfoResponse brand = 2;
    CategoryInfoResponse category = 3;
}

// 商品
message GoodsFilterRequest {
    sint32 page = 1;
    sint32 limit = 2;
    string keywords = 3;
    sint32 is_hot = 4;
    sint32 is_hew = 5;
    sint32 price_min = 6;
    sint32 price_max = 7;
    sint32 brand_id = 8;
    sint32 category_id = 9;
}

message GoodsListResponse {
    int32 total = 1;
    repeated GoodsDetailResponse data = 2;
}

message BatchGoodsByIdRequest {
    repeated GoodsRequest data = 1;
}

message GoodsByIdRequest {
    sint32 id = 1;
}

message GoodsDetailResponse {
    sint32 id = 1;
    sint32 category_id = 2;
    sint32 brand_id = 3;
    string name = 4;
    string goods_sn = 5;
    string subtitle = 6;
    float market_price = 7;
    float shop_price = 8;
    sint32 sold_num = 9;
    sint32 click_num = 10;
    sint32 fav_num = 11;
    bool on_sale = 12;
    bool is_new = 13;
    bool is_hot = 14;
    bool ship_free = 15;
    string front_image = 16;
    repeated string images = 17;
    repeated string desc_images = 18;
}

message GoodsRequest {
    sint32 id = 1;
    string name = 2;
    string goods_sn = 3;
    string subtitle = 4;
    sint32 category_id = 5;
    sint32 brand_id = 6;
    float market_price = 7;
    float shop_price = 8;
    bool ship_free = 9;
    bool on_sale = 10;
    bool is_new = 11;
    bool is_hot = 12;
    sint32 click_num = 13;
    sint32 fav_num = 14;
    string front_image = 15;
    repeated string images = 16;
    repeated string desc_images = 17;
}